### PR TITLE
Export a bundle of files needed to bootstrap a second chef-server

### LINKED ISSUE #566: chef-automate bootstrap export command

### PR DESCRIPTION
**Pull request by yzl**
### :nut_and_bolt: Description
Add API call and command to export a bundle of files needed to bootstrap an A2 chef-server deployment on another node.

Fixes #566.

TODO
- [x] Change command to `chef-automate bootstrap bundle create` for parity with the airgap command.

### :white_check_mark: Checklist

- [ ] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [x] Code actually executed?
- [ ] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/automate-cli/cmd/chef-automate/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		Hidden: true,
+	}
+
+	var bundleCmd = &cobra.Command{
+		Use:    "bundle COMMAND",
+		Hidden: true,
+	}
+
+	var createCmd = &cobra.Command{
+		Use: "create <current working dir>/bootstrap_bundle.tar",
```
Comment (jaym) [inline components/automate-cli/cmd/chef-automate/bootstrap.go]: should we just let people tell us where they want the file written to?

File: components/automate-cli/cmd/chef-automate/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	bootstrapCmd.AddCommand(bundleCmd)
+	bundleCmd.AddCommand(createCmd)
+
+	return bootstrapCmd
+}
+
+func runBootstrapBundleCreate(cmd *cobra.Command, args []string) error {
+	connection, err := client.Connection(client.DefaultClientTimeout)
+	if err != nil {
+		fmt.Println("Connection failed")
```
Comment (jaym) [inline components/automate-cli/cmd/chef-automate/bootstrap.go]: do we need to do this? I thought the stuff we wrap the CLI with handles printing out the error

File: components/automate-cli/cmd/chef-automate/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		return err
+	}
+
+	stream, err := connection.BootstrapBundle(context.Background(), &api.BootstrapBundleRequest{})
```
Comment (jaym) [inline components/automate-cli/cmd/chef-automate/bootstrap.go]: TODO for some point: we probably want this to timeout

File: components/automate-cli/cmd/chef-automate/bootstrap.go (hunk 1 above)
Comment (jaym) [reply components/automate-cli/cmd/chef-automate/bootstrap.go]: this is what `airgap bundle create` looks like:
```
chef-automate airgap bundle create [/path/to/bundle.aib] [flags]
```

File: components/automate-cli/cmd/chef-automate/bootstrap.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		data := resp.GetData()
+		if len(data) <= 0 {
+			continue
+		}
+		if _, err := w.Write(data); err != nil {
+			return status.Annotate(err, status.FileAccessError)
+		}
+	}
+
+	w.Flush()
```
Comment (jaym) [inline components/automate-cli/cmd/chef-automate/bootstrap.go]: error checking?

File: components/automate-cli/cmd/chef-automate/bootstrap.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return status.Annotate(err, status.FileAccessError)
+	}
+	downloadedBundlePath := filepath.Join(cwd, "bootstrap-bundle.tar")
+	downloadedBundleFile, err := os.OpenFile(downloadedBundlePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return status.Annotate(err, status.FileAccessError)
+	}
+
+	defer downloadedBundleFile.Close()
```
Comment (jaym) [inline components/automate-cli/cmd/chef-automate/bootstrap.go]: i think the linter will complain about this

File: components/automate-deployment/pkg/bootstrap/bundle_creator.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+	"strings"
+
+	"github.com/chef/automate/components/automate-deployment/pkg/services"
+	"github.com/chef/automate/lib/io/fileutils"
+	"github.com/chef/automate/lib/product"
+	"github.com/pkg/errors"
+)
+
+var (
+	rootUname   = "root"
```
Comment (jaym) [inline components/automate-deployment/pkg/bootstrap/bundle_creator.go]: There must be a better way to describe this mapping

File: components/automate-deployment/pkg/server/bootstrap_bundle.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/pkg/errors"
+)
+
+// BootstrapBundle makes and downloads a bootstrap bundle
+func (s *server) BootstrapBundle(req *api.BootstrapBundleRequest, stream api.Deployment_BootstrapBundleServer) error {
+	var b bytes.Buffer
+	tarWriter := bufio.NewWriter(&b)
+	bundleCreator := bootstrap.NewBundleCreator()
+
+	pkgs := make([]string, 0)
```
Comment (jaym) [inline components/automate-deployment/pkg/server/bootstrap_bundle.go]: [nit] we know the length of `s.deployment.ExpectedServices`

File: components/automate-deployment/pkg/server/bootstrap_bundle.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	for _, e := range s.deployment.ExpectedServices {
+		pkgs = append(pkgs, e.Name())
+	}
+	err := bundleCreator.Create(pkgs, tarWriter)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create the bootstrap bundle.")
+	}
+
+	tarWriter.Flush()
```
Comment (jaym) [inline components/automate-deployment/pkg/server/bootstrap_bundle.go]: error checking?

Comment (jaym) [inline components/automate-deployment/pkg/server/bootstrap_bundle.go]: error checking?

File: lib/product/types.go [hunk 9]
```diff
@@ -121,6 +122,35 @@ func (c *Collection) validate(allowedPackageSet map[PackageName]bool) error {
...
+func (b *BootstrapSpec) validate() error {
+	switch b.Type {
+	case BootstrapTypeFile:
+		if b.Path == "" {
+			return errors.Errorf("Must provide a path file type %s", BootstrapTypeFile)
+		}
+		if filepath.IsAbs(b.Path) {
+			return errors.Errorf("path must be a relative path inside the services /hab/svc/svc-name directory")
+		}
+		// TODO: validate that the path stays in the service directory
```
Comment (stevendanna) [inline lib/product/types.go]: I assume here you are thinking about paths like `../../../..`?

File: components/automate-deployment/pkg/bootstrap/bundle_creator.go [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+	if err := os.Chmod(absPath, hdr.FileInfo().Mode()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *BundleCreator) mkdir(tarReader *tar.Reader, hdr *tar.Header) error {
+	absPath := path.Join(b.rootDir, hdr.Name)
+	// TODO (jaym): make atomic?
+	err := os.Mkdir(absPath, hdr.FileInfo().Mode())
```
[the remark this answers is not in the document]
Comment (jaym) [reply components/automate-deployment/pkg/bootstrap/bundle_creator.go]: nope. I want the hab root directory to exist as this code does not deal with preserving its permissions. I'll add a comment there

File: components/automate-deployment/pkg/bootstrap/bundle_creator.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+		}
+		groups[gname] = gid
+	}
+
+	exists, err := fileutils.PathExists(b.rootDir)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return errors.Errorf("cannot unpack into %q because it does not", b.rootDir)
```
[the remark this answers is not in the document]
Comment (jaym) [reply components/automate-deployment/pkg/bootstrap/bundle_creator.go]: but it doesn't or isn't

File: components/automate-deployment/pkg/bootstrap/bundle_creator.go [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+		if err != nil {
+			return nil, err
+		}
+
+		header.Name = relPath
+
+		if uid, err := stringutils.IndexOf(b.allowedUsers, header.Uname); err != nil {
+			return nil, errors.Errorf("%q is not an allowed user", header.Uname)
+		} else {
+			header.Uid = uid
```
[the remark this answers is not in the document]
Comment (jaym) [reply components/automate-deployment/pkg/bootstrap/bundle_creator.go]: we were using it, but i changed it to use user. I mostly don't want anything ever trying to use the UID/GID of the system, so I wanted it changed

File: lib/product/types.go (hunk 9 above)
Comment (jaym) [reply lib/product/types.go]: yes

**Pull request merged**